binary_search_tree/tree.py

```diff
@@ -7,51 +7,152 @@ def __init__(self, key, val = None):
+        else:
+            current_node.right = self._add(current_node.right, key, value)
+        return current_node
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        current = self.root
+        while current:
+            if current.key == key:
+                return current.value
+            elif key > current.key:
+                current = current.right
+            else:
+                current = current.left
+        return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        values =[]
+        
+        if self.root == None:
+            return values
+
+        return self._in_order(self.root, values)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def _in_order(self, node, values):
+        if node == None:
+            return 
+
+        self._in_order(node.left, values)
+        values.append({
+            'key': node.key,
+            'value': node.value
+        })
+        self._in_order(node.right, values)
+
+        return values
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def preorder(self):
-        pass
+        values = []
+
+        if self.root == None:
+            return values
+
+        return self._pre_order(self.root, values)
+
+    def _pre_order(self, node, values):
+        if node == None:
+            return
 
-    # Time Complexity: 
-    # Space Complexity:     
+        values.append({
+            'key': node.key,
+            'value': node.value
+        })
+        self._pre_order(node.left, values)
+        self._pre_order(node.right, values)
+        
+        return values
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def postorder(self):
-        pass
+        values = []
+
+        if self.root == None:
+            return values
+
+        return self._post_order(self.root, values)
+
+    def _post_order(self, node, values):
+        if node == None:
+            return 
+
+        self._post_order(node.left, values)
+        self._post_order(node.right, values)
+        values.append({
+            'key': node.key,
+            'value': node.value
+        })
 
-    # Time Complexity: 
-    # Space Complexity:     
+        return values
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(log n) 
     def height(self):
-        pass
+        if self.root == None:
+            return 0
+
+        return self._height(self.root)
+
+    def _height(self, node):
+        if node == None:
+            return 0
 
+        l = self._height(node.left)
+        r = self._height(node.right)
+
+        return (1 + max(l, r))
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n)
+#   # Space Complexity: O(n)
     def bfs(self):
```

Comment:
👍  However since you're using an array as a queue the `queue.pop(0)` is an O(n) operation by itself making your BFS solution O(n^2).  You can reduce this by using a deque or a linked list.